orderer/common/follower/follower_chain_test.go

```diff
@@ -83,26 +94,321 @@ func TestFollowerNewChain(t *testing.T) {
+	})
+
+	t.Run("no need to pull", func(t *testing.T) {
+		setup()
+		clusterConsenter.IsChannelMemberReturns(false, nil)
+		resultChain.fill(joinNum + 1) //No gap between the ledger and the join block
+		assert.True(t, joinBlockAppRaft.Header.Number < mockSupport.Height())
+
+		chain, err := follower.NewChain(
+			mockSupport,
+			clusterConsenter,
+			joinBlockAppRaft,
+			options,
+			pullerCreator.CreateBlockPuller,
+			chainCreator.CreateChain,
+			followerCreator.CreateFollower,
+			cryptoProvider)
+		assert.NoError(t, err)
+
+		cRel, status := chain.StatusReport()
+		require.Equal(t, types.ClusterRelationFollower, cRel)
+		require.Equal(t, types.StatusActive, status)
+
+		assert.NotPanics(t, chain.Start)
+		wgFollower.Wait()
+		assert.NotPanics(t, chain.Halt)
+		assert.False(t, chain.IsRunning())
+
+		cRel, status = chain.StatusReport()
+		assert.Equal(t, types.ClusterRelationFollower, cRel)
+		assert.Equal(t, types.StatusActive, status)
+
+		assert.Equal(t, 1, pullerCreator.CreateBlockPullerCallCount())
+		assert.Equal(t, 0, mockSupport.AppendCallCount())
+
+		assert.Equal(t, 0, chainCreator.CreateChainCallCount())
+		assert.Equal(t, 1, followerCreator.CreateFollowerCallCount())
+	})
+	t.Run("overcome pull failures", func(t *testing.T) {
+		setup()
+		clusterConsenter.IsChannelMemberReturns(true, nil)
+
+		failPull := 10
+		pullerCreator = &mocks.BlockPullerCreator{}
+		puller = &mocks.ChannelPuller{}
+		puller.PullBlockCalls(func(i uint64) *common.Block {
+			if i%2 == 1 && failPull > 0 {
+				failPull = failPull - 1
+				return nil
+			}
+			failPull = 10
+			return blockchain.Block(i)
+		})
+		pullerCreator.CreateBlockPullerReturns(puller, nil)
+		options.TimeAfter = timeAfterCount.After
+
+		chain, err := follower.NewChain(
+			mockSupport,
+			clusterConsenter,
+			joinBlockAppRaft,
+			options,
+			pullerCreator.CreateBlockPuller,
+			chainCreator.CreateChain,
+			followerCreator.CreateFollower,
+			cryptoProvider)
+		assert.NoError(t, err)
+
+		cRel, status := chain.StatusReport()
+		assert.Equal(t, types.ClusterRelationMember, cRel)
+		assert.Equal(t, types.StatusOnBoarding, status)
+
+		assert.NotPanics(t, chain.Start)
+		wgChain.Wait()
+		assert.NotPanics(t, chain.Halt)
+		assert.False(t, chain.IsRunning())
+
+		cRel, status = chain.StatusReport()
+		assert.Equal(t, types.ClusterRelationMember, cRel)
+		assert.Equal(t, types.StatusActive, status)
+
+		assert.Equal(t, 52, pullerCreator.CreateBlockPullerCallCount())
+		assert.Equal(t, 11, mockSupport.AppendCallCount())
+		for i := uint64(0); i <= joinNum; i++ {
+			assert.Equal(t, blockchain.Block(i).Header, resultChain.Block(i).Header, "failed block i=%d", i)
+		}
+		assert.Equal(t, 1, chainCreator.CreateChainCallCount())
+		assert.Equal(t, 0, followerCreator.CreateFollowerCallCount())
+
+		assert.Equal(t, 50, timeAfterCount.getInvocations())
+		assert.Equal(t, int64(5000), timeAfterCount.getMaxDelay())
+	})
+}
+
+type memoryBlockChain struct {
```

Comment:
For this type, and the one below, can we not have counterfeiter generate stubs, and then bind specific behavior on top? Not a huge savings in code, but some, and it will handle locking for us, give us call counts, etc.

Reply:
I got rid of the `timeAfterCounter` for a counterfeiter mock.

With the `memoryBlockChain` I want to keep it because it gets slightly more complicated in the next PR, and making it mock only will make the code unreadable, I think. Lets revisit that then, if we see it can be converted to purely a mock operated thing I'll do it then.

Note that `memoryBlockChain`  is operated by mocks:  the counterfeiter mocks call two instances of `memoryBlockChain`:

The ledger resources mock calls the instance (`resultChain`) modeling the local blockchain:
```go
		mockSupport = &mocks.LedgerResources{}
		mockSupport.ChannelIDReturns("my-channel")
		mockSupport.HeightCalls(resultChain.Height)
		mockSupport.BlockCalls(resultChain.Block)
		mockSupport.AppendCalls(func(block *common.Block) error {
			resultChain.Append(block)
			return nil
		})
```

whereas the block puller mock calls an instance (`blockchain`) that models the remote blockchain we're pulling from:
```go
	pullerCreator = &mocks.BlockPullerCreator{}
		puller = &mocks.ChannelPuller{}
		puller.PullBlockCalls(func(i uint64) *common.Block { return blockchain.Block(i) })
		pullerCreator.CreateBlockPullerReturns(puller, nil)
```
(lines 150-180)

I renamed them `localBlockchain` and `remoteBlockchain` respectively.

Reply:
Sure, if this makes more sense in a later PR, we can defer and discuss later if necessary.